### PR TITLE
Parallel

### PR DESCRIPTION
**Pull request by nielsdebruin**
This pipeline is currently rather slow to remove bottlenecks. 

- Currently, _all_ projects are being written to the output file each time a project has completed causes a substantial delay when the number of function grows. 
- Some projects with a large number of files take the amount of time to be processed yet since everything is sequential they form a bottleneck. This PR allows the pipeline to become parallel making it a lot faster.
- For every project, a data frame is created (just like before) and write to disk to avoid race conditions. 
- Since all of these dataframes have a similar layout they can simply be concatenated with 1 line of codes after the pipeline has completed.
- An option has been added for the number of parallel jobs
- Lastly, a progress bar that is compatible with parallel processing is added such that the rate/total progress can be monitored.
- Some of the statistics have been removed from the pipeline however they can also all be recovered from the dataframes. Only failed file count is not possible to be recovered since it is not written currently. 
- Lastly I added caching and option to change the output dir

### CODE REVIEW

**Comment by alangerak** [comment]
Could you add `joblib` to requirements.txt ?

**Comment by nielsdebruin** [comment]
> Could you add `joblib` to requirements.txt ?

Done

File: .gitignore [hunk 1]
```diff
@@ -106,3 +106,5 @@ venv.bak/
 # project specific
 /projects
 /output
+
+.idea/
```
Comment (casperboone) [inline .gitignore]: You can do this in a global gitconfig, no need to do it here :)

File: pipeline.py [hunk 2]
```diff
@@ -46,101 +50,97 @@ def read_file(filename: str) -> str:
...
+
+        print(f'Extracting for {project_id}...')
+        extracted_functions = {}
+        for filename in list_files(filtered_project_directory):
+            try:
+                functions = extractor.extract(read_file(filename))
+                extracted_functions[filename] = functions
+            except ParseError:
+                print(f"Could not parse file {filename}")
+                # statistics['unparsabl_files'] += 1
```
Comment (casperboone) [inline pipeline.py]: Can you remove this line (or add a new way of collecting statistics)

File: utils.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+def ParallelExecutor(use_bar='tqdm', **joblib_args):
+    def aprun(bar=use_bar, **tq_args):
+        def tmp(op_iter):
+            if str(bar) in all_bar_funcs.keys():
+                bar_func = all_bar_funcs[str(bar)](tq_args)
+            else:
+                raise ValueError("Value %s not supported as bar type"%bar)
+            return Parallel(**joblib_args)(bar_func(op_iter))
+        return tmp
+    return aprun
```
Comment (casperboone) [inline utils.py]: New line :)

**Pull request merged**